### ISSUE TITLE
AArch64: MarkUnsafeAccessTest assert(deoptee.is_compiled_frame()) failed: Wrong frame type

### ISSUE THREAD

**Issue by sanzinger** [oracle/graal#314]
Unittest MarkUnsaveAccessTest.testCompiled fails on AArch64 with following assertion:

```
# To suppress the following error report, specify this argument
# after -XX: or in .hotspotrc:  SuppressErrorAt=/deoptimization.cpp:184
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/sanzinger/graal/jdk/jdk11/src/hotspot/share/runtime/deoptimization.cpp:184), pid=6880, tid=6881
#  assert(deoptee.is_compiled_frame()) failed: Wrong frame type
#
# JRE version: OpenJDK Runtime Environment (11.0) (fastdebug build 11-internal+0-adhoc.sanzinger.jdk11)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 11-internal+0-adhoc.sanzinger.jdk11, compiled mode, tiered, jvmci, compressed oops, g1 gc, linux-aarch64)
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/sanzinger/graal/graal/compiler/hs_err_pid6880.log
OpenJDK 64-Bit Server VM warning: increase O_BUFLEN in ostream.hpp -- output truncated
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
```
Stacktrace:
```
---------------  T H R E A D  ---------------

Current thread (0x0000ffffb401c800):  JavaThread "main" [_thread_in_Java, id=11883, stack(0x0000ffffb8b02000,0x0000ffffb8d02000)]

Stack: [0x0000ffffb8b02000,0x0000ffffb8d02000],  sp=0x0000ffffb8cfdc70,  free space=2031k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x112a8b8]  VMError::report_and_die(int, char const*, char const*, std::__va_list, Thread*, unsigned char*, void*, void*, char const*, int, unsigned long)+0x14c
V  [libjvm.so+0x112b720]  VMError::report_and_die(Thread*, char const*, int, char const*, char const*, std::__va_list)+0x50
V  [libjvm.so+0x7883c0]  report_vm_error(char const*, int, char const*, char const*, ...)+0xe0
V  [libjvm.so+0x7afc6c]  Deoptimization::fetch_unroll_info_helper(JavaThread*, int)+0x1944
V  [libjvm.so+0x7b04d4]  Deoptimization::fetch_unroll_info(JavaThread*, int)+0xbc
v  ~DeoptimizationBlob
v  ~RuntimeStub::load_mirror_patching Runtime1 stub
C  0x0000ffffb401c800

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
v  ~DeoptimizationBlob
v  ~RuntimeStub::load_mirror_patching Runtime1 stub
V  [libjvm.so+0xdfa144]  Monitor::lock_without_safepoint_check()+0x54
C  0x0000000000001ee0

[error occurred during error reporting (printing Java stack), id 0xe0000000]
```

Command to reproduce this issue on JDK9 and JDK11: `mx -v unittest -Xcomp -XX:TieredStopAtLevel=3  MarkUnsafeAccessTest#testCompiled`. 
/cc @adinn

**Comment by adinn** [comment]
Hi Stefan,

I am looking into this. It may actually relate to work Andrew Haley is
doing so I will co-ordinate with him to be sure it is neither 1) a
result of changes he has made or 2) a problem he has already fixed.

regards,


Andrew Dinn
-----------
Senior Principal Software Engineer
Red Hat UK Ltd
Registered in England and Wales under Company Registration No. 03798903
Directors: Michael Cunningham, Michael ("Mike") O'Neill, Eric Shander

**Comment by adinn** [comment]
On 14/03/18 10:28, Andrew Dinn wrote:
> I am looking into this. It may actually relate to work Andrew Haley is
> doing so I will co-ordinate with him to be sure it is neither 1) a
> result of changes he has made or 2) a problem he has already fixed.

It might well do, yes.  I need to find a way to make my current source
available.

-- 
Andrew Haley
Java Platform Lead Engineer
Red Hat UK Ltd. <https://www.redhat.com>
EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

**Comment by adinn** [comment]
Hi Stefan

Sorry for the delay - I have had some other tasks I needed to complete.

When I run this in gdb I hit a SIGBUS when the test program tries to read the unmapped byte buffer memory and after continuing the test completes successfully. I can only see the problem you report in gdb if I configure it as follows
  (gdb) handle SIGBUS nostop noprint pass
At that point I can look at the frame for Deoptimization::fetch_unroll_info_helper and inspect the current frame and its ancestors.

The entry to Deoptimization::fetch_unroll_info is happening indirectly under a runtime stub named "load_mirror_patching Runtime1 stub" (I found this out by unwinding the stack by eyeball).

Meanwhile, the code which tries to unroll the calling stack frames is trying to locate a compiled Java frame or chain of such Java frames somewhere above the runtime deopt call. It has worked its way up the stack looking for a compiled Java top frame, only it has clearly unwound the frames incorrectly since it has finished with an externalVFrame for C++ method Monitor::lock_without_safepoint_check.

I am pretty sure that this error is something to do with in-progress changes made by Andrew Haley. So, until he can push those changes into jdk/hs I think this is just going to break. I am hoping that he will be able to confirm this.
